### ISSUE TITLE
Numeric#nonzero? considered harmful?

### ISSUE THREAD

**Issue by BigDanG** [rubocop-hq/ruby-style-guide#711]
It appears to that `Numeric#nonzero?` is an inconsistent, unconventional predicate method in that it has a {`nil`, `self`} return value range as opposed to the conventional {`false`, `true`}. Should there be a stylistic restriction on its use, especially in predicate statements? I've been bitten by this gotcha, assuming it was intuitively just a negation of the `Numeric#zero?` predicate. This gotcha manifested when I was doing an equality check between predicates in my code and got cruel reminder of the `Numeric#nonzero?` semantics via a propagated `nil`.

**Comment by tom-lord** [comment]
> as opposed to the conventional {`false`, `true`}

This is not a convention. It's more common, sure. But there is nothing *wrong* with the `Numeric#nonzero?` method.

Predicate methods in ruby should - by convention - return either a `truthy` or `falsey` value, and act only as queries - i.e. not alter any underlying state.

For example, [`File#size?`](https://ruby-doc.org/core/File.html#method-c-size-3F) returns either the size (an `Integer`) if the file exists, or `nil` if the object does not exist.

> I was doing an equality check between predicates

Suppose you have two "boolean" variables, `x` and `y`. (But note that ruby, really, *everything* is a boolean!!) You want to check that "both are truthy", or "both are falsey". Rather than something error prone, since it relies on the assumption that `x` and `y` are both `true` or `false`, like:

```
x == y
```

You could instead do one of the following:

```
(x && y) || (!x && !y)
!(x ^ y)
!!x == !!y
```

Personally, I prefer the third variation above. However, this is currently not considered styleguide compliant - although that's been [a subject of debate for a long time](https://github.com/rubocop-hq/ruby-style-guide/issues/591).

**Comment by BigDanG** [comment]
It seems to be that predicates fall into two return value range categories:
- {`true`, `false`}
- {*query object*, `nil`}

You're right that the strict convention is just Ruby truthiness semantics, but I would consider the {`true`, `false`} case to be typical and most in line with the Principle of Least Surprise. And I would consider it highly unconventional if something mixed those 2 forms above, like if `File#size?` had return value range {*file size*, `false`}, even though it adheres to truthiness semantics.

The thing I feel that's wrong about `Numeric#nonzero?` is that it seems to especially violate the Principle of Least Surprise because there's a corresponding `Numeric#zero?` that follows the {`true`, `false`} return value range. I don't think it would be unreasonable for someone to assume that `!a.zero?` is equivalent to `a.nonzero?` without prior knowledge of the gotcha in the spec.

Because of Ruby's truthy/falsy gotchas, and the somewhat inelegant workaround expressions you've presented, the language seems to have a want of a truth-equivalence/truth-equality operator to enforce safe logical equivalence tests. It seems Matz hit on the same idea when he proposed (and subsequently withdrew) [this feature](https://bugs.ruby-lang.org/issues/13067) to add such a concept to the `===` operator on `TrueClass` and `FalseClass`. I'd probably be happier if such an operator, say, defined as `=?`, existed so that the hypothetical expression `!a.zero? =? a.nonzero?` held true.

**Comment by tom-lord** [comment]
I agree with you that the behaviour of `Numberic#nonzero?` vs `Numberic#zero?` is inconsistent/surprising. This has been [the subject of debate for a long time](https://bugs.ruby-lang.org/issues/9123).

The history of why the language ended up in its current design, and proposed changes - as discussed in that thread - is quite interesting; I recommend reading through it.

Personally, I think the best solution would be for ruby add a new method:

```
class Object
  def non(*args)
    self unless args.include? self
  end
end
```

And change the behaviour of `Numeric#nonzero?` to be a `true`/`false` return type. Then, if you _really_ want to `Integer`/`nil` return value, you could use:

```
integer.non(0)
```

...Anyway, I don't think it's rubocop's job to say "don't use a non-deprecated method in the core library". I think this issue should be taken up with the core team - for example, in that thread - and perhaps changed in the next major version release.

**Comment by anthonyms** [comment]
It may not always be possible to replace `x.nonzero?` with `!x.zero?` as in the following case.

`[1,0,1,0].map(&:nonzero?)`

**Comment by jlamberg** [comment]
> It may not always be possible to replace `x.nonzero?` with `!x.zero?` as in the following case.
> 
> `[1,0,1,0].map(&:nonzero?)`

@anthonyms, unless you use a lambda, for example:

```
[1,0,1,0].map(&->(item){ !item.zero? })
=> [true, false, true, false]
```

or 

```
nonzero = ->(item){ !item.zero? }
[1,0,1,0].map(&nonzero)
=> [true, false, true, false]
```

**Comment by marcandre** [comment]
The origin (and correct usecase) for `nonzero?` is to map `0` to a falsy value, e.g.:

```ruby
def <=>(other)
  return unless other.is_a?(MyClass)

  (foo <=> other.foo).nonzero? ||
    (bar <=> other.bar).nonzero? ||
    (baz <=> other.baz)
end
```

Maybe it should have been called `nonzero`. But today this is how it is named, and how it should be used.
Note that `defined?` also does not return `true/false`, nor [does `String < Hash`](https://bugs.ruby-lang.org/issues/9123#note-25) or any of `Module#autoload? / File.world_readable? / world_writable? / size? / Encoding.compatible?`

I'm am :-1: on that rule.

**Comment by rockaBe** [comment]
> It may not always be possible to replace `x.nonzero?` with `!x.zero?` as in the following case.
> 
> `[1,0,1,0].map(&:nonzero?)`

Or the explicit block syntax:
```ruby
[1, 0, 1, 0].map { |item| !item.zero? }
```

**Comment by marcandre** [comment]
> > It may not always be possible to replace `x.nonzero?` with `!x.zero?` as in the following case.
> > `[1,0,1,0].map(&:nonzero?)`
> 
> Or the explicit block syntax:
> 
> ```ruby
> [1, 0, 1, 0].map { |item| !item.zero? }
> ```

@rockaBe These are not equivalent.

**Comment by pirj** [comment]
Closing as controversial. There is no agreement to deprecate the method on Ruby Core side as well.

**Comment by gregschmit** [comment]
Sorry if commenting on long-closed issues is considered impolite, but I wanted to mention that this bit me today.

A Rails project I'm working on had a boolean column called `unpaid`, and records were being created like this:

```
EventRegistration.create!(
  **registration_params,
  user: current_user,
  unpaid: event.price.nonzero?,  # This is obviously always `true` or `false`.
)
```

However we noticed many records where `unpaid` was `nil`, because `EventRegistration.where(unpaid: false)` was not returning all the paid registrations. Well, there are multiple ways those events are created so I decided to make that column `NOT NULL` so it would only contain `true` and `false` values. Then, the above event registration `create!` call started failing, because `event.price.nonzero?` was returning `nil` rather than `false`.

So I fixed it with `unpaid: !event.price.zero?` (though `unpaid: !!event.price.nonzero?` would have been fine, too).

I would love to use a lint rule that forbids using this predicate method.
